### PR TITLE
Add DeepOmit for hash and array

### PR DESCRIPTION
**Pull request by lyzhao**
## Description

The PR adds new functionality to enable the omission of a field recursively in an object and an array. Please refer to: https://stackoverflow.com/questions/55539387/deep-omit-with-typescript

Example: 
```
type Input =  {
    __typename: string,
    a: string,
    nested: {
        __typename: string,
        b: string
    }
    nestedArray: Array<{
        __typename: string,
        b: string
    }>
    nestedTuple: [{
        __typename: string,
        b: string
    }]
}

type InputWithoutKey = DeepOmit<Input, '__typename'>

let s: InputWithoutKey = {
    a: "",
    nested: {
        b:""
    },
    nestedArray: [
        {b: ""}
    ],
    nestedTuple: [
        { b: ""},
    ]
}
```

## Checklist

- [x] I have added a changelog entry, prefixed by the type of change noted above (Documentation fix and Test update does not need a changelog as we do not publish new version)

### CODE REVIEW

File: packages/useful-types/src/types.ts [hunk 1]
```diff
@@ -79,3 +79,29 @@ export type ConstructorArgumentAtIndex<
...
+  | Symbol
+  | undefined
+  | null;
+type DeepOmitHelper<T, K extends keyof T> = {
+  [P in K]: T[P] extends infer TP
+    ? TP extends Primitive
+      ? TP
+      : TP extends any[]
+        ? DeepOmitArray<TP, K>
+        : DeepOmit<TP, K>
```
[the remark this answers is not in the document]
Comment (lyzhao) [reply packages/useful-types/src/types.ts:98]: What do you think? Should we go with the linter?

[the remark this answers is not in the document]
Comment (atesgoral) [reply packages/useful-types/src/types.ts:98]: Can't merge without making the linter happy, but I didn't like the linter's formatting suggestion :/

[the remark this answers is not in the document]
Comment (lyzhao) [reply packages/useful-types/src/types.ts:98]: We could update the linter as a separate PR. That will reformat this and other parts of the code.

[the remark this answers is not in the document]
Comment (lyzhao) [reply packages/useful-types/src/types.ts:98]: @atesgoral bump

[the remark this answers is not in the document]
Comment (atesgoral) [reply packages/useful-types/src/types.ts:98]: Good lord. https://github.com/prettier/prettier/issues/5814

[the remark this answers is not in the document]
Comment (atesgoral) [reply packages/useful-types/src/types.ts:98]: Let's just with the weird flat style Prettier enforces (while I vehemently disagree) and move on :)

**Review by atesgoral — APPROVED** [review]

**Comment by atesgoral** [comment]
@lyzhao `@shopify/useful-types` version 2.4.0 has been released

**Pull request merged**